xds/src/main/java/io/grpc/xds/XdsServerWrapper.java

```diff
@@ -0,0 +1,441 @@
+  }
+
+  @Override
+  public Server start() throws IOException {
+    checkState(started.compareAndSet(false, true), "Already started");
+    syncContext.execute(new Runnable() {
+      @Override
+      public void run() {
+        internalStart();
+      }
+    });
+    Exception exception;
+    try {
+      exception = initialStartFuture.get();
+    } catch (InterruptedException | ExecutionException e) {
+      throw new RuntimeException(e);
+    }
+    if (exception != null) {
+      throw (exception instanceof IOException) ? (IOException) exception :
+              new IOException(exception);
+    }
+    return this;
+  }
+
+  private void internalStart() {
+    try {
+      xdsClientPool = xdsClientPoolFactory.getOrCreate();
+    } catch (Exception e) {
+      StatusException statusException = Status.UNAVAILABLE.withDescription(
+              "Failed to initialize xDS").withCause(e).asException();
+      listener.onNotServing(statusException);
+      initialStartFuture.set(statusException);
+      return;
+    }
+    xdsClient = xdsClientPool.getObject();
+    // TODO(chengyuanzhang): add an API on XdsClient indicating if it is using v3, don't get
+    //  from bootstrap.
+    boolean useProtocolV3 = xdsClient.getBootstrapInfo().getServers().get(0).isUseProtocolV3();
+    String listenerTemplate = xdsClient.getBootstrapInfo().getServerListenerResourceNameTemplate();
+    if (!useProtocolV3 || listenerTemplate == null) {
+      StatusException statusException =
+          Status.UNAVAILABLE.withDescription(
+              "Can only support xDS v3 with listener resource name template").asException();
+      listener.onNotServing(statusException);
+      initialStartFuture.set(statusException);
+      xdsClient = xdsClientPool.returnObject(xdsClient);
+      return;
+    }
+    discoveryState = new DiscoveryState(listenerTemplate.replaceAll("%s", listenerAddress));
+  }
+
+  @Override
+  public Server shutdown() {
+    if (!shutdown.compareAndSet(false, true)) {
+      return this;
+    }
+    syncContext.execute(new Runnable() {
+      @Override
+      public void run() {
+        internalShutdown();
+        delegate.shutdown();
+      }
+    });
+    syncContext.drain();
+    return this;
+  }
+
+  @Override
+  public Server shutdownNow() {
+    if (!shutdown.compareAndSet(false, true)) {
+      return this;
+    }
+    syncContext.execute(new Runnable() {
+      @Override
+      public void run() {
+        internalShutdown();
+        delegate.shutdownNow();
+      }
+    });
+    syncContext.drain();
+    return this;
+  }
+
+  // Must run in SynchronizationContext
+  private void internalShutdown() {
+    logger.log(Level.FINER, "Shutting down XdsServerWrapper");
+    if (discoveryState != null) {
+      discoveryState.shutdown();
+    }
+    if (xdsClient != null) {
+      xdsClient = xdsClientPool.returnObject(xdsClient);
+    }
+    if (timeService != null) {
+      SharedResourceHolder.release(GrpcUtil.TIMER_SERVICE, timeService);
```

Comment:
You should do this:
```
timeService = SharedResourceHolder.release(GrpcUtil.TIMER_SERVICE, timeService);
```

otherwise `timeService` is never set to `null`

Reply:
This is still not done. Is that because of the other changes made? Not doing this causes an exception when I ran a unit test

Reply:
ah, we shouldn't be using timerservice once it is released. I moved cancel before release. Not doing assignment to null because we wanted it to be final. It is a tradeoff.

> Not doing this causes an exception when I ran a unit test

guess its because of using it after release?

Reply:
It's an existing unit test (in this PR) and running it in the debugger showed the exception (which was suppressed). Ack about timeService being final